quickwit/quickwit-search/src/find_trace_ids_collector.rs

```diff
@@ -0,0 +1,508 @@
+        self.span_timestamp
+            .cmp(&other.span_timestamp)
+            .reverse()
+            .then(self.term_ord.cmp(&other.term_ord))
+    }
+}
+
+impl PartialOrd for TraceIdTermOrd {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for TraceIdTermOrd {
+    fn eq(&self, other: &Self) -> bool {
+        self.term_ord == other.term_ord
+    }
+}
+
+impl Eq for TraceIdTermOrd {}
+
+/// Finds the most recent trace ids among a set of matching spans. Multiple spans belonging to the
+/// same trace can be found in the document set. As a result, this problem is akin to finding the
+/// top k elements with duplicates
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct FindTraceIdsCollector {
+    /// The number of traces to select.
+    pub num_traces: usize,
+    /// The name of the fast field storing the trace IDs.
+    pub trace_id_field_name: String,
+    /// The name of the fast field recording the spans' start timestamp.
+    pub span_timestamp_field_name: String,
+}
+
+impl FindTraceIdsCollector {
+    /// The names of the fast fields accessed by this collector.
+    pub fn fast_field_names(&self) -> HashSet<String> {
+        HashSet::from_iter([
+            self.trace_id_field_name.clone(),
+            self.span_timestamp_field_name.clone(),
+        ])
+    }
+
+    /// The field names of the term dictionnaries accessed by this collector.
+    pub fn term_dict_field_names(&self) -> HashSet<String> {
+        HashSet::from_iter([self.trace_id_field_name.clone()])
+    }
+}
+
+impl Collector for FindTraceIdsCollector {
+    type Fruit = Vec<TraceIdFruit>;
+    type Child = FindTraceIdsSegmentCollector;
+
+    fn for_segment(
+        &self,
+        _segment_local_id: u32,
+        segment_reader: &SegmentReader,
+    ) -> tantivy::Result<Self::Child> {
+        let trace_id_ff_reader = segment_reader
+            .fast_fields()
+            .u64s(&self.trace_id_field_name)?;
+        let span_timestamp_column = segment_reader
+            .fast_fields()
+            .date(&self.span_timestamp_field_name)?;
+        let trace_id_field = segment_reader
+            .schema()
+            .get_field(&self.trace_id_field_name)?;
+        let inverted_index_reader = segment_reader.inverted_index(trace_id_field)?;
+        Ok(Self::Child {
+            trace_id_ff_reader,
+            span_timestamp_column,
+            inverted_index_reader,
+            select_trace_ids: SelectTraceIds::new(self.num_traces),
+        })
+    }
+
+    fn merge_fruits(
+        &self,
+        segment_fruits: Vec<<Self::Child as SegmentCollector>::Fruit>,
+    ) -> tantivy::Result<Self::Fruit> {
+        let mut trace_ids = Vec::with_capacity(self.num_traces);
+        let mut seen_trace_ids = FnvHashSet::default();
+
+        for trace_id in segment_fruits
+            .into_iter()
+            .map(BinaryHeap::from)
+            .map(binary_heap::IntoIterSorted::new)
```

Comment:
Why do we need these two lines?

It would be less complicated and faster to just sort the segment fruits, and apply the kmerge on that.